Allow to override Ltac2 mutable definitions at runtime (#13982)
coq/coq: doc/sphinx/proof-engine/ltac2.rst

      the `f` in the body is resolved dynamically. This is witnessed by
      the second re-definition.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer: I agree, and it's even more confusing that it looks just like a vernacular command.

> .. tacn:: Set @qualid := @ltac2_expr

Author: This term syntax is similar to the corresponding vernacular command. There is the same parallelism with Definitions and let, so... Also, `Set` is a reserved keyword, so it can't be used as a constructor name in Ltac2.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 2: You mean similar to this, right?

![image](https://user-images.githubusercontent.com/1253341/112934712-35358380-90d7-11eb-96a3-f8677c3cf30d.png)

I don't get the point of your statement "Also, `Set` is a reserved keyword, so it can't be used as a constructor name in Ltac2."  Why does that matter?

This doesn't seem worth breaking the lower-case convention for tactics, which all built-in tactics follow.

> .. tacn:: Set @qualid := @ltac2_expr

Author: > You mean similar to this, right?

Yes.

> Why does that matter?

This is not a built-in tactic, it is a term node (for typing reasons). It needs to use a dedicated syntax that is specific enough to be 1. clearly distinct from said tactics and 2. not introducing crazy parsing conflicts. Set is a keyword, there is thus no possibility of rule clashing. I don't really care about the actual syntax used, *as long as it minimizes potential conflicts*. In particular, that means that I won't accept a syntax using `set` instead. (It would break the `set` tactic immediately.)

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 3: What about something like `with_...` in the spirit of the `with_strategy` tactical? I agree with the concern over breaking standard casing convention.

> .. tacn:: Set @qualid := @ltac2_expr

Author: That's not a keyword. Also, this is a global state so it is not well-scoped.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 4: Would it help with parsing conflicts if the syntax were given in user-land via `Ltac2 Notation`?  (Is there any reason the syntax can't be given that way?)

If it's possible to have parsing factor, I'd be inclined to just use `ident := expr`.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 4: Or we could follow Ltac1 and use `ident ::= expr`.  That's currently unused in Ltac2, right?

> .. tacn:: Set @qualid := @ltac2_expr

Author: > Would it help with parsing conflicts 

Probably not, since that means the syntax is too generic. We precisely want to have a syntax that cannot clash with user-defined notations.

> ident := expr

This one is already used for mutable *fields*.

> ident ::= expr

It's unused in Ltac2, but this is an infix notation without a leading keyword, so expect trouble. Also, it's the kind of stuff users would typically define with a notation, so see point 1.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 4: > We precisely want to have a syntax that cannot clash with user-defined notations.

Why is this?  I'm of the opposite opinion: there should be as little as possible in the way of magic features that have to be hard-coded.  If it were possible for all syntax to be user-defined, I'd prefer that world.

> .. tacn:: Set @qualid := @ltac2_expr

Author: Ltac2 already features side-effects, in particular a global state. This PR has been asked by quite a few people already.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 5: Why use this separate system instead of regular `ref`?

> .. tacn:: Set @qualid := @ltac2_expr

Author: > there should be as little as possible in the way of magic features that have to be hard-coded.

@JasonGross I agree with the highest confidence for *primitive tactics*. But this is *not* a primitive tactic. It *is* literal built-in magic that needs to be handle specifically by the type-checker and the runtime system. Just as you can't define let-bindings or function abstraction with tactic notations, you shouldn't be able to define runtime modification of mutable globals either.

> .. tacn:: Set @qualid := @ltac2_expr

Author: @SkySkimmer both typing and runtime.
- You can't reliably store mutable data across summaries, which is why Ltac2 globals are restricted to values. Also, value restriction would force you to resolve weak polymorphic type variables before proceeding further, so it'd be quite a mess. It has other nice consequences like the fact that some operations are required to be effectful like environment access which prevents you to keep pointer to stale ressources.
- At the typing level, you need to ensure that you're not messing around with capture of polymorphic variables.

The restriction solves both problems at once but you need runtime support for global replacement, which is precisely what this PR introduces.

> .. tacn:: Set @qualid := @ltac2_expr

Reviewer 4: > Just as you can't define let-bindings or function abstraction with tactic notations, you shouldn't be able to define runtime modification of mutable globals either.

You can't define the *semantics* in user-space (and I don't object to using the FFI for implementing the primitive).  But the syntax can be defined in user-space.  For example, here is function abstraction syntax defined in user-space:
https://github.com/coq/coq/blob/6effcc263beded0d530d724fab8edae86815adf8/theories/Unicode/Utf8_core.v#L31-L34
Certainly you need some primitive way of building the construct, and if you're saying that this global-set primitive is so special that it can't even be given a type in Ltac2, and has extra special typing rules associated with it, then I'll withdraw my objection.  (Though in fact I would instead prefer extending the typing rules of Ltac2 to support the primitive.)  But if you can do something like
```coq
Ltac2 @ external set_global : 'a global_mutable_ident -> 'a -> unit := "ltac2" "set_global".
```
then I don't see any reason to hardcode the primitive into the AST.

Perhaps what you're saying, though, is that Ltac2 types don't have a notion of contexts, and so you can't define a well-typed/well-scoped syntax for function abstraction?  (Perhaps I'm wishing that Ltac2 notations were more like hygenic macros (perhaps a la Lean) and less like mixfix (a la Agda) + interpretation scopes?)

>    This expression redefines a previous ``mutable`` definition at runtime.
>    Contrarily to its command equivalent above, it allows to replace a definition
>    from inside a tactic, and will take effect when the expression is evaluated.
>    This construct has type :n:`unit`, and the class of definitions allowed to be
>    replaced at runtime is currently restricted to monomorphic ones. It does not
>    feature the :n:`as` clause of the command, since the same effect can be
>    obtained using a simple let-binding.

Reviewer 2: ```suggestion
   Redefines a previous ``mutable`` definition at runtime.
   Unlike its command equivalent above, it allows replacing a definition
   from inside a tactic.  The redefinition takes effect when the expression is evaluated.
   This construct has type :n:`unit`.  Currently only monomorphic definitions can be
   replaced at runtime. It does not
   feature the :n:`as` clause of the command, since the same effect can be
   obtained using a simple let-binding.
```
